File: backend/app/schemas/recurring.py
"""Pydantic schemas for recurring rules (REC-01..REC-05)."""

import uuid
from datetime import date

from pydantic import BaseModel, ConfigDict, Field

from app.models.recurring import RecurringFrequency
from app.models.transactions import TransactionType
from app.schemas.transactions import Money


class RecurringRuleBase(BaseModel):
    type: TransactionType
    amount: Money
    amount_is_estimate: bool = False
    currency: str = Field(pattern="^[A-Z]{3}$")
    description: str = Field(min_length=1, max_length=200)
    category_id: uuid.UUID | None = None
    payment_method_id: uuid.UUID | None = None
    frequency: RecurringFrequency
    start_date: date
    end_date: date | None = None
    max_occurrences: int | None = Field(default=None, ge=1)
    month_day: int | None = Field(default=None, ge=1, le=31)
    use_last_day: bool = False


class RecurringRuleCreate(RecurringRuleBase):
    pass


class RecurringRuleUpdate(BaseModel):
    """REC-04: edits only affect future instances."""

    amount: Money | None = None
    amount_is_estimate: bool | None = None
    description: str | None = Field(default=None, min_length=1, max_length=200)
    category_id: uuid.UUID | None = None
    payment_method_id: uuid.UUID | None = None
    end_date: date | None = None
    max_occurrences: int | None = Field(default=None, ge=1)
    month_day: int | None = Field(default=None, ge=1, le=31)
    use_last_day: bool | None = None
    is_active: bool | None = None


class RecurringRuleOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    type: TransactionType
    amount: Money
    amount_is_estimate: bool
    currency: str
    description: str
    category_id: uuid.UUID | None
    payment_method_id: uuid.UUID | None
    frequency: RecurringFrequency
    start_date: date
    end_date: date | None
    max_occurrences: int | None
    month_day: int | None
    use_last_day: bool
    is_active: bool

File: frontend/src/api/recurring.ts
// Reglas recurrentes (REC-01..05).
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { api } from "./client";

export interface RecurringRuleOut {
  id: string;
  type: "expense" | "income";
  amount: string;
  amount_is_estimate: boolean;
  currency: string;
  description: string;
  category_id: string | null;
  payment_method_id: string | null;
  frequency: "weekly" | "biweekly" | "monthly" | "yearly";
  start_date: string;
  end_date: string | null;
  max_occurrences: number | null;
  month_day: number | null;
  use_last_day: boolean;
  is_active: boolean;
}

export interface RecurringRuleBody {
  type: string;
  amount: string;
  currency: string;
  description: string;
  category_id?: string | null;
  payment_method_id?: string | null;
  frequency: string;
  start_date: string;
  month_day?: number | null;
  use_last_day?: boolean;
  amount_is_estimate?: boolean;
}

export interface RecurringRuleUpdate {
  id: string;
  amount?: string;
  description?: string;
  category_id?: string | null;
  payment_method_id?: string | null;
  end_date?: string | null;
  max_occurrences?: number | null;
  is_active?: boolean;
}

export function useRecurringRules(includeInactive = false) {
  return useQuery({
    queryKey: ["recurring-rules", includeInactive],
    queryFn: () =>
      api<RecurringRuleOut[]>(
        `/api/v1/recurring-rules?include_inactive=${includeInactive}`,
      ),
  });
}

function useInvalidateRecurring() {
  const qc = useQueryClient();
  // El pronóstico proyecta nómina/domiciliados desde las reglas (PRO-03/04).
  return () =>
    ["recurring-rules", "forecast"].forEach(
      (k) => void qc.invalidateQueries({ queryKey: [k] }),
    );
}

export function useCreateRecurringRule() {
  const invalidate = useInvalidateRecurring();
  return useMutation({
    mutationFn: (body: RecurringRuleBody) =>
      api<RecurringRuleOut>("/api/v1/recurring-rules", {
        method: "POST",
        body: JSON.stringify(body),
      }),
    onSuccess: invalidate,
  });
}

export function useUpdateRecurringRule() {
  const invalidate = useInvalidateRecurring();
  return useMutation({
    mutationFn: ({ id, ...body }: RecurringRuleUpdate) =>
      api<RecurringRuleOut>(`/api/v1/recurring-rules/${id}`, {
        method: "PATCH",
        body: JSON.stringify(body),
      }),
    onSuccess: invalidate,
  });
}

export function useDeleteRecurringRule() {
  const invalidate = useInvalidateRecurring();
  return useMutation({
    mutationFn: (id: string) =>
      api<void>(`/api/v1/recurring-rules/${id}`, { method: "DELETE" }),
    onSuccess: invalidate,
  });
}
